File: Step 03 Arrays/Step 3.3 Hard/LongestSubarrayZeroSum/LongestSubarrayZeroSum.java


class GfG
{
    int maxLen(int arr[], int n)
    {
        // Your code here
        int max_len = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        int sum = 0;
        map.put(0, -1);
        for (int i=0; i<n; ++i) {
            sum += arr[i];
            if (map.containsKey(sum) || sum==0) {
                max_len = Math.max(max_len, (i - map.get(sum)));
            } else {
                map.put(sum, i);
            }
        }
        return max_len;
    }
}

File: Step 03 Arrays/Step 3.3 Hard/LongestSubarrayZeroSum/LongestSubarrayZeroSum.cpp
/*You are required to complete this function*/

class Solution {
  public:
    int maxLen(vector<int>& arr, int n) {
        // Your code here
        int max_len = 0;
        unordered_map<int, int> map;
        int sum = 0;
        map[0] = -1;
        for (int i=0; i<n; ++i) {
            sum += arr[i];
            if (map.find(sum)!=map.end() || sum==0) {
                max_len = max(max_len, (i - map[sum]));
            } else {
                map[sum] = i;
            }
        }
        return max_len;
    }
};